docs/proposal/content-library-support.md

```diff
@@ -0,0 +1,134 @@
+
+* Use ContentLibrary items as the VM image template.
+
+### Non-Goals
+
+* Turn content library the only supported template source.
+
+## Proposal
+
+### User Stories
+
+#### Story 1
+
+As an admin, I'd like to manage templates/OVAs across the vCenters that can be consumed by CAPV.
+
+#### Story 2
+
+As an cluster user, I'd like to use content library items(templates/OVAs) as the source of cluster nodes.
+
+### Overall Design
+
+The design focuses on how to clone VMs from content library items. For the content library management itself users can reference vSphere documentation.
+
+### Implementation Details/Notes/Constraints
+
+#### Naming scheme
+
+We use `Template` field of `VirtualMachineCloneSpec` to specify the source we want to clone,
+since we want to augment it to also from the content library, we need to have a way to differentiate the source.
+I propose we use URI scheme for that, it has the format: `scheme://location` which
+
+* scheme, can be vsphere|library
+  * vsphere for conventional vm templates
+  * library for OVA/templates from content library
+* to make it backward compatible, the scheme part can be omitted for conventional vm templates, so if we encounter
+  template without scheme part, we assume it is vsphere template, existing CAPV does not need to do anything.
+
+#### API changes
+
+##### VirtualMachineCloneSpec CR
+
+The `VirtualMachineCloneSpec` has following changes, the spec is used to specify clone spec for `VsphereVM`, it needs the following change to deal with newly added content library capabilities.
+
+```golang
+type VirtualMachineCloneSpec struct {
+  // Template is the name or inventory path of the template used to clone
+  // the virtual machine. It can be a vSphere template, or a content library item. For example:
+  // vsphere template: [vsphere://location of the vsphere template], vsphere:// can be omitted for backward compatibility
+  // content library: [library://name of the content library item]
+  // note that this has impact on cloneMode field, since it is only possible to do full clone from content library items,
+  // and we should have validation rule for this.
+  // +kubebuilder:validation:MinLength=1
+  Template string `json:"template"`
+
+  // The secret that stores the credentials for accessing the content library.
+  // It should reside on the same namespace of the Cluster that reference it.
+  // It is optional since enabling access control for the content library is optional. It's up to content library admin to decide if
+  // it is required to enable access control.
+  // +optional
+  LibraryCredentials string `json:"library_credentials"`
+}
+```
+
+#### Implementation Details
+
+##### VSphereVM clone changes
+
+We need to enhance the [clone](https://github.com/kubernetes-sigs/cluster-api-provider-vsphere/blob/19c5deb79e97a8cd9d546ec39db3f29231bbcb91/pkg/services/govmomi/vcenter/clone.go#L75) function to recogize the scheme we defined, and choose appropriate API to clone the VM.
```

Comment:
I'd like some more details on this before proceeding. I think you will want to use https://github.com/vmware/govmomi/blob/main/vapi/vcenter/vcenter_ovf.go#L292-L306 and then https://github.com/vmware/govmomi/blob/252e157f27b33d5cebeecdb21212e3f3fbe44039/object/virtual_machine.go#L72 

We already have this logic in VM Service mode at https://github.com/vmware-tanzu/vm-operator/blob/main/pkg/vmprovider/providers/vsphere/session/session_vm_create.go

Reply:
Also some good direct reference in https://github.com/vmware/govmomi/blob/main/govc/library/deploy.go

Reply:
I believe it needs to download the image content from Content Library when deploying.
We need to be careful here as it may run multiple times and create a heavy workload to vCenter if we need multiple VM instances.
I think it would be better to import the image as a separate template and do the clone as what we have today.

Reply:
When creating a subscribed content library, there are two options for dealing with the content, download immediately/when needed. 
I think we should advise user to use eager option(download immediately).
If user choose lazy option(when needed), I think the work round here is to deploy a test vm before creating clusters, or create a small test cluster to fetch the library items to local.

Reply:
The problem with download immediately is that we may end up subscribing to CL with a lot of images that may be problematic for long term storage space.

Anyway, can we agree this is not an enhancement question but more an implementation detail?